gordon/metrics/log.py

```diff
@@ -0,0 +1,173 @@
+class LogTimer:
+    """Timer that outputs metrics to the logger.
+
+    Args:
+        metric (dict): Metric to output.
+        logger (.LoggerAdapter): Object used for outputting metrics.
+        time_unit (str): Unit for LogTimer output.
+            Options: s, ms, ns.
+    """
+    def __init__(self, metric, logger, time_unit):
+        self.metric = metric
+        self.logger = logger
+        self.time_multiplier = self._time_unit_to_multiplier(time_unit)
+
+    def _time_unit_to_multiplier(self, unit):
+        return {
+            's': 1,
+            'ms': 1E3,
+            'ns': 1E9
+        }[unit]
+
+    async def __aenter__(self):
+        await self.start()
+        return self
+
+    async def __aexit__(self, *args):
+        await self.stop()
+
+    async def start(self):
+        """Start the timer."""
+        self._start_time = time.perf_counter()
+
+    async def stop(self):
+        """Stop the timer and output the time delta to the logger."""
+        time_elapsed = (
+            time.perf_counter() - self._start_time) * self.time_multiplier
+        self.metric['value'] = time_elapsed
+        self.logger.log(self.metric)
+
+
+class LoggerAdapter:
+    """Configurable adapter that formats and outputs metrics.
+
+    Args:
+        level (str): Log level at which to output metric.
+    """
+    LOGFMT = '[{metric_name}] value: {value}'
+
+    def __init__(self, level):
+        self.level = level
+        self._logger = logging.getLogger('gordon.metrics-logger')
+
+    def log(self, metric):
+        """Format and output metric.
+
+        Args:
+            metric (dict): Complete metric.
+        """
+        message = self.LOGFMT.format(**metric)
+        if metric['context']:
+            message += ' context: {context}'.format(context=metric['context'])
+        getattr(self._logger, self.level)(message)
+
+
+@zope.interface.implementer(interfaces.IMetricRelay)
+class LogRelay:
+    """Manage metrics that get output to a logger.
+
+    Args:
+        config (dict): Required keys:
+            level (str): Log level of metrics log messages.
+            time_unit (str): Unit for LogTimer output.
+                Options: s, ms, ns. Defaults to 's'.
+    """
+    def __init__(self, config):
+        level = config.get('log_level', 'info')
+        self.time_unit = config.get('time_unit', 's')
+        self.logger = LoggerAdapter(level)
+        self.counters = collections.defaultdict(int)
+
+    def _create_metric(self, metric_name, value, context, **kwargs):
+        context = context or {}
+
+        return {
+            'metric_name': metric_name,
+            'value': value,
+            'context': context
+        }
+
+    async def incr(self, metric_name, value=1, context=None, **kwargs):
+        """Increase a metric counter by `value` and output result.
+
+        Args:
+            metric_name (str): Identifier of the metric.
+            value (int): (optional) Value with which to increase the metric.
```

Comment:
It's not really an `int` though right? Could be a float?

Reply:
I changed it to `number` since that's a broader type.